lib/mix/lib/mix/tasks/test/stale.ex

```diff
@@ -0,0 +1,261 @@
+defmodule Mix.Tasks.Test.Stale do
+  @moduledoc false
+
+  require Mix.Compilers.Elixir
+  alias Mix.Compilers.Elixir, as: CE
+
+  import Record
+
+  defrecordp :source, [
+    source: nil,
+    compile_references: [],
+    runtime_references: [],
+    external: []
+  ]
+
+  @stale_manifest ".compile.test_stale"
+  @manifest_vsn :v1
+
+
+  @doc """
+  Set up the `--stale` run for given matched test files, test paths, and options.
+
+  If there are tests to run, a tuple containing the test files to run, the PID
+  of the agent used in the each_module callback, and the callback options for
+  ParallelRequire is returned.
+
+  If there are no tests to run, the pid and callbacks are not returned, since the
+  test task should not perform a ParallelRequire if there are no tests to run.
+
+  The test task is responsible for running the returned tests, using the returned
+  ParallelRequire callbacks, calling the agent_write_manifest/1 function with the
+  provided PID after performing the ParallelRequire sucessfully, and calling the
+  agent_stop/1 function with the same PID after the ParallelRequire, whether it
+  succeeds or not.
+  """
+  def set_up(matched_test_files, test_paths, opts) do
+    manifest = manifest()
+    modified = Mix.Utils.last_modified(manifest)
+    all_sources = read_manifest()
+
+    removed =
+      for source(source: source) <- all_sources,
+          not(source in matched_test_files),
+          do: source
+
+    configs = Mix.Project.config_files
+    force = opts[:force] || Mix.Utils.stale?(configs, [manifest]) || test_helper_stale?(test_paths)
```

Comment:
Beautiful. You remembered of the config files as well as the test helper. :D I believe you should also include the elixir manifest though.